tests/unit/basket/test_offers.py

```diff
@@ -0,0 +1,162 @@
+Selector, UK = get_classes("partner.strategy", ["Selector", "UK"])
+
+
+class UKSelector(Selector):
+    @staticmethod
+    def strategy(request=None, user=None, **kwargs):
+        return UK(request)
+
+
+def money(amount):
+    return amount.quantize(D("0.01"), ROUND_HALF_UP)
+
+
+def get_user_basket(user, request):
+    editable_baskets = Basket.objects.filter(status__in=["Open", "Saved"])
+    basket, __ = editable_baskets.get_or_create(owner=user)
+    basket.strategy = UKSelector().strategy(request=request, user=user)
+    basket.reset_offer_applications()
+    if not basket.is_empty:
+        Applicator().apply(basket, user, request)
+    request.session[settings.OSCAR_BASKET_COOKIE_OPEN] = basket.pk
+    request.session.save()
+    return basket
+
+
+class OfferTest(TestCase):
+    fixtures = ["catalogue", "auth", "offer"]
+
+    def check_general_truths(self, basket):
+        self.assertEqual(
+            basket.total_discount,
+            basket.total_incl_tax_excl_discounts - basket.total_incl_tax,
+            "The total discount given should be equal to the price without "
+            "discount minus the price with the discount applied",
+        )
+        inverse_tax_multiplier = D(1) / (D(1) + UK.rate)
+        calculated_total_excl_tax = money(
+            inverse_tax_multiplier * basket.total_incl_tax
+        )
+
+        self.assertEqual(
+            calculated_total_excl_tax,
+            basket.total_excl_tax,
+            "The total price without tax should conform to the standard "
+            "formula for calculating tax (as a percentage)",
+        )
+        self.assertAlmostEqual(
+            basket.total_excl_tax_excl_discounts / basket.total_incl_tax_excl_discounts,
+            basket.total_excl_tax / basket.total_incl_tax,
+            4,
+            "The ratio of price with tax and without tax should be the same for the "
+            "price with and without discounts. If that is not the case people would "
+            "be able to change the tax they must pay by gaming the discount.",
+        )
+        self.assertNotAlmostEqual(
+            basket.total_excl_tax_excl_discounts - basket.total_excl_tax,
+            basket.total_incl_tax_excl_discounts - basket.total_incl_tax,
+            2,
+            "The discount over the total excluding tax can never be the same as "
+            "the discount over the total including tax. Otherwise our tax rate"
+            "would not be linear over the amount.",
+        )
+
+    def test_offer_incl_tax(self):
+        "The offer should be calculated as if it was declared including tax"
+        with self.settings(OSCAR_OFFERS_INCL_TAX=True):
+            self.assertEqual(Basket.objects.count(), 0)
+
+            admin = User.objects.get(username="admin@admin.nl")
+            self.client.force_login(admin)
+
+            # throw an item in the basket
+            basket_add_url = reverse("basket:add", args=(2,))
```

Comment:
Rather than faking all this request logic, it will be much easier to user `oscar.test.factories.create_basket` and then manipulate the basket with `oscar.test.basket.add_product`.